src/_guides/libraries/objective-c-interop.md

```diff
@@ -293,6 +293,44 @@ then check the status, and wait for the duration of the audio file:
     }
 ```
 
+### Callbacks and multithreading limitations
+
+While `ffigen` does support converting
+Dart functions to Objective-C blocks,
+there are some limitations to keep in mind
+when using them.
+
+1. Dart isolates are not the same thing as threads.
+   An isolate is not guaranteed to run on a particular thread,
+   and the VM may change which thread an isolate is running on
+   without warning.
+   There is an [open feature request][] to allow isolates to be
+   pinned to specific theads.
+2. Most Apple APIs don't have any guarantees about
+   which thread a callback will be run on.
+3. Many Apple APIs can only be called from the main thread,
+   also known in Flutter as the platform thread.
+   This includes any APIs that involve UI interaction.
+
+Points 1 and 2 mean that a callback created in one isolate
+may be invoked on a thread running a different isolate,
+or no isolate at all.
+This will cause your app to crash.
```

Comment:
Will it crash, or just it might crash?

Reply:
@dcharkes I think there's an assert in the FFI trampoline that checks for a current isolate, right?

Reply:
Correct.

runtime/vm/runtime_entry.cc

```c++
// Not registered as a runtime entry because we can't use Thread to look it up.
static Thread* GetThreadForNativeCallback(uword callback_id,
                                          uword return_address) {
  Thread* const thread = Thread::Current();
  if (thread == nullptr) {
    FATAL("Cannot invoke native callback outside an isolate.");
  }
  if (thread->no_callback_scope_depth() != 0) {
    FATAL("Cannot invoke native callback when API callbacks are prohibited.");
  }
  if (thread->is_unwind_in_progress()) {
    FATAL("Cannot invoke native callback while unwind error propagates.");
  }
  if (!thread->IsMutatorThread()) {
    FATAL("Native callbacks must be invoked on the mutator thread.");
  }
```

Reply:
But right before that it says "may be invoked on a thread running a different isolate, or no isolate at all". The former won't crash, it will just have unexpected behavior, right?

Reply:
Later in `GetThreadForNativeCallback` there's a check that the isolate is correct:

```c++
thread->VerifyCallbackIsolate(callback_id, return_address);
```

Reply:
Sounds like this statement is correct for both scenarios, can this be resolved @liamappelbe?

Reply:
I think so.